### PR TITLE
MAGN-6761 Reload camera position for background preview from file.

### PR DESCRIPTION
**Pull request by ikeough**
### Purpose

This PR relates to [MAGN-6761](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6761). This feature adds the ability to save the orientation of the background camera to the .dyn file. It does this by exposing two workspace events that enable the view to hook the file saving/loading process. These new events are `WorkspaceModel.Saving` and `DynamoModel.WorkspaceLoading`. An architectural review is recommended as these events allow access to the `XmlDocument` representing the .dyn during its creation. Misuse of these events could cause corruption of the saved file.

### Impact assessment

- [ ] Changes the way 3rd party nodes work (Add ikeough to the reviewer list)
- [ ] Includes migrations (Add ikeough to the reviewer list)
- [x] Changes a public interface (Add lukechurch to the reviewer list)
- [ ] Affects which components of the system depends on which other components (Add pboyer to the reviewer list)
- [ ] Introduces new components (Add lukechurch to the reviewer list)
- [ ] Has a performance impact (Add keyu to the reviewer list)
- [ ] Changes the installer (Add sharadkjaiswal to the reviewer list)

#### Testing strategy

- Automated tests are available in `CameraTests.cs`.
- Manual testing should cover the following scenarios.
  - Dynamo opened. Camera is oriented to default position.
  - No camera data in .dyn, file loads successfully and camera is oriented to default position.
  - Good camera data in .dyn file. File loads successfully and camera orientation is set correctly.
  - Corrupt camera data in .dyn file. File loads successfully. Camera orientation is set to default. Warning message is displayed in the console.
  - File saved and reloaded. Camera is oriented as before saving.
  - Workspace cleared. Camera is set to default orientation.

### Review staging

- [ ] Architectural Design Review
- [ ] UI Review

### Reviewers

@lukechurch 

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### CODE REVIEW

**Comment by ikeough** [comment]
@lukechurch ping!

**Comment by lukechurch** [comment]
Apologies @ikeough

File: src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs [hunk 1]
```diff
@@ -59,8 +65,11 @@ private void NotifyPropertyChanged(string propertyName)
         private double lightAzimuthDegrees = 45.0;
         private double lightElevationDegrees = 35.0;
         private int renderingTier;
+        private DynamoViewModel viewModel;
         private const double NearPlaneDistanceFactor = 0.0001;
-
+        internal Vector3D defaultCameraLookDirection = new Vector3D(-10, -10, -10);
```
Comment (lukechurch) [inline src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs]: Can these be made read only?

File: src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs [hunk 2]
```diff
@@ -927,6 +1059,44 @@ private void TestSelection(object parameters)
...
+                var lx = float.Parse(cameraNode.Attributes["lookX"].Value);
+                var ly = float.Parse(cameraNode.Attributes["lookY"].Value);
+                var lz = float.Parse(cameraNode.Attributes["lookZ"].Value);
+
+                Camera.LookDirection = new Vector3D(lx, ly, lz);
+                Camera.Position = new Point3D(ex, ey, ez);
+            }
+            catch (Exception ex)
+            {
+                viewModel.Model.Logger.Log("Camera data could not be loaded.", LogLevel.Console);
```
Comment (lukechurch) [inline src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs]: Also log the exception please

**Comment by lukechurch** [comment]
Overall lgtm - Can you test what happens if the code that hooks into the saving pipeline misbehaves? We need to make sure that Dynamo crashes appropriately whilst warning the user that their work has not been saved.

Lets also test that the code that walks over the graph for serialisation to send to instrumentation behaves correctly - and gets the information including the camera position.

**Comment by ikeough** [comment]
@lukechurch I wrapped the save logic in a try/catch. The save file is still valid if the camera data isn't written.

I moved the raising of the `Workspace.Saving` event inside `PopulateXmlDocument` so that it's picked up in the standard save path and also when `GetStringRepOfWorkspace` is called for analytics. I verified that the cameras are coming through correctly in both scenarios.

**Comment by lukechurch** [comment]
Thanks @ikeough  LGTM

**Pull request merged**